File: 04_quicksort/java/04_recursive_max.java
import java.util.*;
import java.lang.*;
import java.io.*;
import java.util.ArrayList;

class grokkingAlgorithms
{	

	public static int maxx(ArrayList<Integer> arr){
		if(arr.size() == 2){
			return arr.get(0) > arr.get(1) ? arr.get(0) : arr.get(1);
		}
		int temp = arr.get(0);
		arr.remove(0);
		int result = maxx(arr);
		return temp > result ? temp : result;
	}
	
	public static void main (String[] args) throws java.lang.Exception
	{
		ArrayList<Integer> arr = new ArrayList<Integer>();
		arr.add(1);
		arr.add(2);
		arr.add(3);
		arr.add(4);
		System.out.print(maxx(arr)); 
	}
}

File: 04_quicksort/c++/04_recursive_max.cpp
#include <iostream>
#include <vector>
using namespace std;

int maxx(vector<int> arr){
	if (arr.size() == 2)
		return arr[0] > arr[1] ? arr[0] : arr[1];
	else{
		int temp = arr[0];
		arr.erase(arr.begin(), arr.begin()+1);
		int submax = maxx(arr);
		return temp > submax ? temp : submax;
	}
}
int main(int argc, char *argv[]) {
	vector<int> arr{1, 2, 3, 4};
	cout << maxx(arr) << endl;
}